utilities/blob_db/blob_compaction_filter.cc

```diff
@@ -360,11 +360,27 @@ CompactionFilter::BlobDecision BlobIndexCompactionFilterGC::PrepareBlobOutput(
 
   PinnableSlice blob;
   CompressionType compression_type = kNoCompression;
+  std::string compression_output;
   if (!ReadBlobFromOldFile(key, blob_index, &blob, false, &compression_type)) {
     gc_stats_.SetError();
     return BlobDecision::kIOError;
   }
 
+  // If the compression_type is changed, re-compress it with the new compression
+  // type.
+  if (compression_type != blob_db_impl->bdb_options_.compression) {
+    if (compression_type != kNoCompression) {
+      const Status status =
+          blob_db_impl->DecompressSlice(blob, compression_type, &blob);
+      if (!status.ok()) {
+        return BlobDecision::kCorruption;
+      }
+    }
```

Comment:
`ReadBlobFromOldFile` already supports decompression; we could simply introduce a boolean `const bool compression_changed = (compression_type != blob_db_impl->bdb_options_.compression;` and pass it as `need_decompress` to avoid having to explicitly call `DecompressSlice` here.

Reply:
Thanks @ltamasi for the review.
For this issue, the problem is the compression_type is also returning from `ReadBlobFromOldFile()`. (I guess it's combining the getting the compression_type and reading data in one call to reduce IO or mutex lock).
One option is to pass a `new_compression_type` into `ReadBlobFromOldFile()`, then only decompress if it's different from the old one. But that feels not a clean function.

Reply:
Right, we have a chicken and egg problem because `compression_type` is set by `ReadBlobFromOldFile` :)